packages/contracts/contracts/plugin/PluginSetupProcessor.sol

```diff
@@ -186,394 +247,394 @@ contract PluginSetupProcessor is DaoAuthorizable {
+        // once again if the plugin was uninstalled and then installed.
+        if (pluginState.blockNumber < pluginState.setupIds[newSetupId]) {
+            revert SetupAlreadyPrepared(setupId);
         }
 
-        emit UpdatePrepared({
-            sender: msg.sender,
-            dao: _dao,
-            pluginSetup: _updateParams.newPluginSetup,
-            data: _data,
-            plugin: _updateParams.plugin,
-            updatedHelpers: updatedHelpers,
-            permissions: permissions,
-            initData: initData
-        });
+        pluginState.setupIds[newSetupId] = block.number;
 
-        return (permissions, initData);
+        // Avoid stack too deep.
+        emitPrepareUpdateEvent(_dao, newSetupId, _params, preparedDependency, initData);
+
+        return (initData, preparedDependency);
     }
 
     /// @notice Applies the permissions of a prepared update of an UUPS upgradeable contract to a DAO.
     /// @param _dao The address of the updating DAO.
-    /// @param _plugin The address of the `PluginUUPSUpgradeable` proxy contract.
-    /// @param _pluginSetup The address of the `PluginSetup` contract.
-    /// @param _pluginSetupRepo The repository storing the `PluginSetup` contracts of all versions of a plugin.
-    /// @param _initData The initialization data to be passed to the upgradeable plugin contract via `upgradeToAndCall`. // revisit
-    /// @param _permissions The list of multi-targeted permission operations to apply to the updating DAO.
-    function applyUpdate(
-        address _dao,
-        address _plugin,
-        address _pluginSetup,
-        PluginRepo _pluginSetupRepo,
-        bytes memory _initData,
-        PermissionLib.ItemMultiTarget[] calldata _permissions
-    ) external canApply(_dao, APPLY_UPDATE_PERMISSION_ID) {
-        bytes32 setupId = _getSetupId(_dao, _pluginSetup, address(_pluginSetupRepo), _plugin);
-
-        if (updatePermissionHashes[setupId] != _getPermissionsHash(_permissions)) {
-            revert PermissionsHashMismatch();
+    /// @param _params The struct containing the parameters for the `applyInstallation` function.
+    function applyUpdate(address _dao, ApplyUpdate calldata _params)
+        external
+        canApply(_dao, APPLY_UPDATE_PERMISSION_ID)
+    {
+        bytes32 pluginId = _getPluginId(_dao, _params.plugin);
+
+        PluginState storage pluginState = states[pluginId];
+
+        bytes32 setupId = _getSetupId(
+            _params.pluginSetupRef,
+            pHash(_params.permissions),
+            _params.helpersHash,
+            _params.initData,
+            PreparationType.Update
+        );
+
+        // If the plugin block number exceeds the setupId preparation block number,
+        // This means applyUpdate was already called on another setupId
+        // and all the rest setupIds should become idle or setupId is not prepared before.
+        if (pluginState.blockNumber >= pluginState.setupIds[setupId]) {
```

Comment:
This check to revert could be moved earlier, since the statement above doesn't use `pluginState`

Reply:
I need `setupId`.

Reply:
Shame on me, was looking at pluginId and not setupId. Resolve